# Syllabus: Applied Statistics for High-Throughput Biology

## Instructor

Levi Waldron, PhD  
Associate Professor of Biostatistics  
City University of New York School Graduate of Public Health and Health Policy  
New York, NY, U.S.A.  

Email: lwaldron.research@gmail.com  
Other contact information: https://waldronlab.io

## Summary

This course will provide biologists and bioinformaticians with practical statistical and data analysis skills to perform rigorous analysis of high-throughput biological data.  The course assumes some familiarity with genomics and with R programming, but does not assume prior statistical training.  It covers the statistical concepts necessary to design experiments and analyze high-dimensional data generated by genomic technologies, including: exploratory data analysis, linear modeling, analysis of categorical variables, principal components analysis, and batch effects.  

## Textbook

* [Biomedical Data Science](https://genomicsclass.github.io/book/) by Irizarry and Love ([ePub version](https://leanpub.com/dataanalysisforthelifesciences/) on Leanpub)
* [Source repository](https://github.com/genomicsclass/labs)
* [Modern Statistics for Modern Biology](https://www.huber.embl.de/msmb/) by Holmes and Huber (secondary)
* [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/release/OSCA/) (OSCA) by Amezquita, Lun, Hicks, Gottardo, Oâ€™Callaghan

## Related Resources

* http://waldronlab.io/teaching/resources/

## Labs

Each day will include a hands-on lab session, that students should attempt in full. 

## Session detail by day

All course materials will be available from https://github.com/waldronlab/AppStatBio/.

1. Introduction
    + random variables
    + distributions
    + hypothesis testing for one or two samples (t-test, Wilcoxon test, etc)
2. Dimensionality reduction
    + Distances in high dimensions
    + Principal Components Analysis and Singular Value Decomposition
    + Multidimensional Scaling
    + t-SNE and UMAP
