<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="AppStatBio">
<title>Applied Statistics for High-throughput Biology: Session 1 • AppStatBio</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Applied Statistics for High-throughput Biology: Session 1">
<meta property="og:description" content="AppStatBio">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">AppStatBio</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/day1_intro.html">Applied Statistics for High-throughput Biology: Session 1</a>
    <a class="dropdown-item" href="../articles/day2_unsupervised.html">Applied Statistics for High-throughput Biology Session 2</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="day1_intro_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><script src="day1_intro_files/viz-1.8.2/viz.js"></script><link href="day1_intro_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="day1_intro_files/grViz-binding-1.0.10/grViz.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Applied Statistics for High-throughput Biology: Session 1</h1>
            
      
      
      <div class="d-none name"><code>day1_intro.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="front-matter">Front matter<a class="anchor" aria-label="anchor" href="#front-matter"></a>
</h2>
<div class="section level3">
<h3 id="welcome">Welcome<a class="anchor" aria-label="anchor" href="#welcome"></a>
</h3>
<p>My research interests:</p>
<ul>
<li>High-dimensional statistics (more variables than observations)</li>
<li>Predictive modeling and methodology for validation</li>
<li>Metagenomic profiling of the human microbiome</li>
<li>Multi-omic data analysis for cancer</li>
<li><a href="https://www.waldronlab.io" class="external-link uri">https://www.waldronlab.io</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="the-textbooks">The textbooks<a class="anchor" aria-label="anchor" href="#the-textbooks"></a>
</h3>
<ul>
<li>
<a href="https://genomicsclass.github.io/book/" class="external-link">Biomedical Data
Science</a> by Irizarry and Love (<a href="https://leanpub.com/dataanalysisforthelifesciences/" class="external-link">ePub
version</a> on Leanpub)</li>
<li><a href="https://github.com/genomicsclass/labs" class="external-link">Source
repository</a></li>
<li>
<a href="https://www.huber.embl.de/msmb/" class="external-link">Modern Statistics for
Modern Biology</a> by Holmes and Huber (secondary)</li>
<li>
<a href="https://bioconductor.org/books/release/OSCA/" class="external-link">Orchestrating
Single-Cell Analysis with Bioconductor</a> (OSCA) by Amezquita, Lun,
Hicks, Gottardo, O’Callaghan</li>
</ul>
<div class="figure">
<img src="https://d2sofvawe08yqg.cloudfront.net/dataanalysisforthelifesciences/hero2x?1549465432" alt=""><p class="caption">Biomedical Data Science book cover</p>
</div>
</div>
<div class="section level3">
<h3 id="day-1-outline">Day 1 outline<a class="anchor" aria-label="anchor" href="#day-1-outline"></a>
</h3>
<ul>
<li>Random variables and distributions</li>
<li>Hypothesis testing for one or two samples (t-test, Wilcoxon test,
etc)</li>
<li>Confidence intervals</li>
<li>Essential data classes in R/Bioconductor</li>
<li>
<a href="https://genomicsclass.github.io/book/" class="external-link">Book</a> chapters 0
and 1</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Define random variables and distinguish them from non-random
ones</li>
<li>Distinguish population from sampling distributions</li>
<li>Identify scenarios for use of z-test and t-tests</li>
<li>Interpret confidence intervals and Standard Error</li>
<li>Identify and interrogate key R/Bioconductor data classes</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="random-variables-and-distributions">Random Variables and Distributions<a class="anchor" aria-label="anchor" href="#random-variables-and-distributions"></a>
</h2>
<div class="section level3">
<h3 id="random-variables">Random Variables<a class="anchor" aria-label="anchor" href="#random-variables"></a>
</h3>
<ul>
<li><p><strong>A random variable</strong>: any characteristic that can
be measured or categorized, and where any particular outcome is
determined at least partially by chance.</p></li>
<li>
<p>Examples:</p>
<ul>
<li>number of new diabetes cases in population in a given year</li>
<li>The weight of a randomly selected individual in a population</li>
</ul>
</li>
<li>
<p>Types:</p>
<ul>
<li>Categorical random variable (e.g. disease / healthy)</li>
<li>Discrete random variable (e.g. sequence read counts)</li>
<li>Continuous random variable (e.g. normalized qPCR intensity)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="random-variables---examples">Random Variables - examples<a class="anchor" aria-label="anchor" href="#random-variables---examples"></a>
</h3>
<p>Normally distributed random variable with mean <span class="math inline">\(\mu = 0\)</span> / standard deviation <span class="math inline">\(\sigma = 1\)</span>, and a sample of <span class="math inline">\(n=100\)</span></p>
<p><img src="day1_intro_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="random-variables---examples-1">Random Variables - examples<a class="anchor" aria-label="anchor" href="#random-variables---examples-1"></a>
</h3>
<p>Poisson distributed random variable (<span class="math inline">\(\lambda = 2\)</span>), and a sample of <span class="math inline">\(n=100\)</span>.</p>
<p><img src="day1_intro_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="random-variables---examples-2">Random Variables - examples<a class="anchor" aria-label="anchor" href="#random-variables---examples-2"></a>
</h3>
<p>Negative Binomially distributed random variable (<span class="math inline">\(size=30, \mu=2\)</span>), and a sample of <span class="math inline">\(n=100\)</span>.</p>
<p><img src="day1_intro_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="random-variables---examples-3">Random Variables - examples<a class="anchor" aria-label="anchor" href="#random-variables---examples-3"></a>
</h3>
<ul>
<li>Binomial Distribution random variable (<span class="math inline">\(size=20, prob=0.25\)</span>), and a sample of
<span class="math inline">\(n=100\)</span>.
<ul>
<li>use for binary outcomes</li>
</ul>
</li>
</ul>
<p><img src="day1_intro_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="hypothesis-testing">Hypothesis Testing<a class="anchor" aria-label="anchor" href="#hypothesis-testing"></a>
</h2>
<div class="section level3">
<h3 id="why-hypothesis-testing">Why hypothesis testing?<a class="anchor" aria-label="anchor" href="#why-hypothesis-testing"></a>
</h3>
<ul>
<li>Allows yes/no statements, e.g. whether:
<ul>
<li>a population mean has a hypothesized value, or</li>
<li>an intervention causes a measurable effect relative to a control
group</li>
</ul>
</li>
<li>Example questions with yes/no answers:
<ul>
<li>Is a gene differentially expressed between two populations?</li>
<li>Do hypertensive, smoking men have the same mean cholesterol level as
the general population?</li>
</ul>
</li>
<li>Hypothesis testing is not the only framework for inferential
statistics, e.g.:
<ul>
<li>confidence intervals</li>
<li>posterior probabilities (Bayesian statistics)</li>
<li>read <a href="http://www.nature.com/news/statistics-p-values-are-just-the-tip-of-the-iceberg-1.17412" class="external-link">p-values
are just the tip of the iceberg</a>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="logic-of-hypothesis-testing">Logic of hypothesis testing<a class="anchor" aria-label="anchor" href="#logic-of-hypothesis-testing"></a>
</h3>
<ul>
<li>Hypotheses are made about <em>populations</em> based on inference
from <em>samples</em>
</li>
<li>We make inference from <em>samples</em> because we usually cannot
observe the entire population</li>
</ul>
<center>
<div class="figure">
<img src="images/population-vs-sample2.jpg" alt=""><p class="caption">Sample vs Population</p>
</div>
</center>
<p>Source: <a href="https://keydifferences.com/wp-content/uploads/2016/06/population-vs-sample2.jpg" class="external-link uri">https://keydifferences.com/wp-content/uploads/2016/06/population-vs-sample2.jpg</a></p>
</div>
<div class="section level3">
<h3 id="one-and-two-sample-hypothesis-tests-of-mean">One and two-sample hypothesis tests of mean<a class="anchor" aria-label="anchor" href="#one-and-two-sample-hypothesis-tests-of-mean"></a>
</h3>
<p><strong>One-sample hypothesis test of mean</strong> - sample comes
from one of two population distributions:</p>
<ol style="list-style-type: decimal">
<li>usual distribution that has been true in the past
<ul>
<li>
<span class="math inline">\(H_0: \mu = \mu_0\)</span> (null
hypothesis)</li>
</ul>
</li>
<li>a potentially new distribution induced by an intervention or
changing condition
<ul>
<li>
<span class="math inline">\(H_A: \mu \neq \mu_0\)</span>
(alternative hypothesis)</li>
</ul>
</li>
</ol>
<p><strong>Two-sample hypothesis test of means</strong> - two samples
are drawn, either:</p>
<ol style="list-style-type: decimal">
<li>from a single population
<ul>
<li>
<span class="math inline">\(H_0: \mu_1 = \mu_2\)</span> (null
hypothesis)</li>
</ul>
</li>
<li>from two different populations
<ul>
<li>
<span class="math inline">\(H_A: \mu_1 \neq \mu_2\)</span>
(alternative hypothesis)</li>
</ul>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="population-vs-sampling-distributions">Population vs sampling distributions<a class="anchor" aria-label="anchor" href="#population-vs-sampling-distributions"></a>
</h3>
<ul>
<li>Population distributions
<ul>
<li>Each realization / point is an individual</li>
</ul>
</li>
<li>Sampling distributions
<ul>
<li>Each realization / point is a sample</li>
<li>distribution depends on sample size</li>
<li>large sample distributions are given by the Central Limit
Theorem</li>
</ul>
</li>
<li>
<strong>Hypothesis testing is about sampling distributions</strong>
<ul>
<li>Did my sample likely come from that distribution?</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="note-about-the-central-limit-theorem">Note about the Central Limit Theorem<a class="anchor" aria-label="anchor" href="#note-about-the-central-limit-theorem"></a>
</h3>
<p>When sample size is large, the average <span class="math inline">\(\bar{Y}\)</span> of a random sample:</p>
<ol style="list-style-type: decimal">
<li>Follows a normal distribution</li>
<li>The normal distribution has mean entered at the population average
<span class="math inline">\(\mu_Y\)</span>
</li>
<li>with standard deviation equal to the population standard deviation
<span class="math inline">\(\sigma_Y\)</span>, divided by the square
root of the sample size <span class="math inline">\(N\)</span>. We refer
to the standard deviation of the distribution of a random variable as
the random variable’s <em>standard error</em>.</li>
</ol>
<p>Thanks to CLT, linear modeling, t-tests, ANOVA are all guaranteed to
work reasonably well for large samples (more than ~30 observations).</p>
<p>Go to online demo: <a href="http://onlinestatbook.com/stat_sim/sampling_dist/" class="external-link uri">http://onlinestatbook.com/stat_sim/sampling_dist/</a></p>
</div>
<div class="section level3">
<h3 id="example-applications-of-hypothesis-tests-for-continuous-variables">Example applications of hypothesis tests for continuous
variables<a class="anchor" aria-label="anchor" href="#example-applications-of-hypothesis-tests-for-continuous-variables"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Do study participants on a diet lose weight compared to a control
group?</li>
<li>Does a gene knockout result in less viable cell colonies, as
measured by the number of living cells in replicate experiments?</li>
<li>Is <em>Prevotella copri</em> more abundant in the guts of vegans
than of meat-eaters?</li>
<li>Do infants born in this region have a higher birthweight than the
national average?</li>
</ol>
<p>These are research hypotheses. What are the corresponding <em>null
hypotheses</em>?</p>
</div>
<div class="section level3">
<h3 id="the-z-tests">The z-tests<a class="anchor" aria-label="anchor" href="#the-z-tests"></a>
</h3>
<ul>
<li>“<span class="math inline">\(z\)</span>” refers to the
<strong>Standard Normal Distribution</strong>: <span class="math inline">\(N(\mu=0, \sigma=1)\)</span>
</li>
<li>In a one-sample test, only a single sample is drawn:
<ul>
<li><span class="math inline">\(H_0: \mu = \mu_0\)</span></li>
</ul>
</li>
<li>In a two-sample test, two samples are drawn <em>independently</em>:
<ul>
<li><span class="math inline">\(H_0: \mu_1 = \mu_2\)</span></li>
</ul>
</li>
<li>A paired test is one sample of paired measurements, e.g.:
<ul>
<li>pain level in individuals before and after treatment
<ul>
<li><span class="math inline">\(H_0: \mu_{before} =
\mu_{after}\)</span></li>
</ul>
</li>
<li>microbial abundance in upper and lower intestines in same
individuals
<ul>
<li><span class="math inline">\(H_0: \mu_{upper} =
\mu_{lower}\)</span></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="when-to-use-z-tests">When to use z-tests<a class="anchor" aria-label="anchor" href="#when-to-use-z-tests"></a>
</h3>
<ul>
<li>
<span class="math inline">\(\frac{{}\bar{x} - \mu}{\sigma}\)</span>
and <span class="math inline">\(\frac{\bar{x_1} -
\bar{x_2}}{\sigma}\)</span> are z-distributed <em>if</em>:
<ul>
<li><strong>the standard deviation <span class="math inline">\(\sigma\)</span> for the population is known in
advance</strong></li>
<li>the population values are normally distributed</li>
<li>the population values are slightly skewed but n &gt; 15</li>
<li>the population values are quite skewed but n &gt; 30</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-of-one-sample-z-test">Example of one-sample z-test<a class="anchor" aria-label="anchor" href="#example-of-one-sample-z-test"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TeachingDemos</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">weights</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span>, mean <span class="op">=</span> <span class="fl">75</span>, sd <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] 68.73546 76.83643 66.64371 90.95281 78.29508 66.79532 79.87429 82.38325</span></span>
<span><span class="co">#&gt;  [9] 80.75781 71.94612</span></span>
<span><span class="fu">TeachingDemos</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TeachingDemos/man/z.test.html" class="external-link">z.test</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">weights</span>,</span>
<span>    mu <span class="op">=</span> <span class="fl">70</span>,  <span class="co">#specified for population under H0 because this is a one-sample test</span></span>
<span>    stdev <span class="op">=</span> <span class="fl">10</span>, <span class="co">#specified for population because this is a z-test</span></span>
<span>    alternative <span class="op">=</span> <span class="st">"two.sided"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  One Sample z-test</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  weights</span></span>
<span><span class="co">#&gt; z = 1.9992, n = 10.0000, Std. Dev. = 10.0000, Std. Dev. of the sample</span></span>
<span><span class="co">#&gt; mean = 3.1623, p-value = 0.04559</span></span>
<span><span class="co">#&gt; alternative hypothesis: true mean is not equal to 70</span></span>
<span><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span><span class="co">#&gt;  70.12408 82.51998</span></span>
<span><span class="co">#&gt; sample estimates:</span></span>
<span><span class="co">#&gt; mean of weights </span></span>
<span><span class="co">#&gt;        76.32203</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-t-tests">The t-tests<a class="anchor" aria-label="anchor" href="#the-t-tests"></a>
</h3>
<ul>
<li>The t-tests are for small samples and do <em>not</em> rely on the
Central Limit Theorem</li>
<li>In a one-sample test, only a single sample is drawn:
<ul>
<li><span class="math inline">\(H_0: \mu = \mu_0\)</span></li>
</ul>
</li>
<li>In a two-sample test, two samples are drawn <em>independently</em>:
<ul>
<li><span class="math inline">\(H_0: \mu_1 = \mu_2\)</span></li>
</ul>
</li>
<li>A paired test is one sample of paired measurements, e.g.:
<ul>
<li>individuals before and after treatment</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="when-to-use-t-tests">When to use t-tests<a class="anchor" aria-label="anchor" href="#when-to-use-t-tests"></a>
</h3>
<ul>
<li>
<span class="math inline">\(\frac{{}\bar{x} - \mu}{s}\)</span> and
<span class="math inline">\(\frac{\bar{x_1} - \bar{x_2}}{s}\)</span> are
t-distributed <em>if</em>:
<ul>
<li><strong>the standard deviation <span class="math inline">\(s\)</span> is estimated from the
sample</strong></li>
<li>the population values are normally distributed</li>
<li>the population values are slightly skewed but n &gt; 15</li>
<li>the population values are quite skewed but n &gt; 30</li>
</ul>
</li>
<li>Wilcoxon tests are an alternative for non-normal populations
<ul>
<li>e.g. rank data</li>
<li>data where ranks are more informative than means</li>
<li>
<em>not</em> when many ranks are arbitrary</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-of-one-sample-t-test">Example of one-sample t-test<a class="anchor" aria-label="anchor" href="#example-of-one-sample-t-test"></a>
</h3>
<p>Note that here we do not specify the standard deviation, because it
is estimated from the sample.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html" class="external-link">t.test</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">weights</span>,</span>
<span>    mu <span class="op">=</span> <span class="fl">70</span>,  <span class="co">#specified for population under H0 because this is a one-sample test</span></span>
<span>    alternative <span class="op">=</span> <span class="st">"two.sided"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  One Sample t-test</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data:  weights</span></span>
<span><span class="co">#&gt; t = 2.5612, df = 9, p-value = 0.03063</span></span>
<span><span class="co">#&gt; alternative hypothesis: true mean is not equal to 70</span></span>
<span><span class="co">#&gt; 95 percent confidence interval:</span></span>
<span><span class="co">#&gt;  70.73805 81.90600</span></span>
<span><span class="co">#&gt; sample estimates:</span></span>
<span><span class="co">#&gt; mean of x </span></span>
<span><span class="co">#&gt;  76.32203</span></span></code></pre></div>
<p>In this particular example the p-value is smaller than from the
<em>z-test</em>, but in general it would be <strong>less
powerful</strong> than a <em>z-test</em> if its assumptions are met.</p>
</div>
</div>
<div class="section level2">
<h2 id="confidence-intervals">Confidence intervals<a class="anchor" aria-label="anchor" href="#confidence-intervals"></a>
</h2>
<div class="section level3">
<h3 id="why-confidence-intervals">Why confidence intervals?<a class="anchor" aria-label="anchor" href="#why-confidence-intervals"></a>
</h3>
<ul>
<li>The above procedures produced both p-values and confidence
intervals</li>
<li>p-values are useful only for deciding whether to reject <span class="math inline">\(H_0\)</span>
</li>
<li>p-values do not report effect size (observed difference).</li>
<li>p-values only indicate statistical significance which does not
guarantee scientific/clinical significance.</li>
</ul>
<p>Confidence intervals provide a <strong>probable range for the true
population mean</strong>.</p>
<p><img src="day1_intro_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="intervals-for-any-confidence-level">Intervals for any confidence level<a class="anchor" aria-label="anchor" href="#intervals-for-any-confidence-level"></a>
</h3>
<ul>
<li>If the sampling distribution is normal with standard error <span class="math inline">\(SE\)</span>, then a confidence interval for the
population mean is:
<ul>
<li>
<span class="math inline">\(\bar{X} \pm 1.96 \times SE\)</span> (95%
CI, normal sampling distribution)</li>
<li>
<span class="math inline">\(\bar{X} \pm z_{\alpha / 2}^{crit} \times
SE\)</span> (normal sampling distribution)</li>
<li>
<span class="math inline">\(\bar{X} \pm t_{\alpha / 2, df}^{crit}
\times SE\)</span> (t sampling distribution)</li>
</ul>
</li>
<li>The part after the <span class="math inline">\(\pm\)</span> is
called the <em>margin of error</em>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="confidence-intervals-vs--hypothesis-testing">Confidence intervals vs. hypothesis testing<a class="anchor" aria-label="anchor" href="#confidence-intervals-vs--hypothesis-testing"></a>
</h3>
<ul>
<li>Confidence intervals can be used for hypothesis testing
<ul>
<li>reject <span class="math inline">\(H_0\)</span> if the “null value”
is not contained in the CI</li>
</ul>
</li>
<li>Do <strong>not</strong> use overlap between two CIs for hypothesis
test
<ul>
<li>for a two sample hypothesis test <span class="math inline">\(H_0:
\mu_1 = \mu_2\)</span>, must construct a single confidence interval for
<span class="math inline">\(\mu_1 - \mu_2\)</span>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="q-a-confidence-intervals">Q &amp; A: confidence intervals<a class="anchor" aria-label="anchor" href="#q-a-confidence-intervals"></a>
</h3>
<p>Which of the following are true?</p>
<ul>
<li>The 95% CI contains 95% of the values in the population</li>
<li>The 95% CI will contain the population mean, 19 times out of 20</li>
<li>The 95% CI is 95% probable to contain the population mean</li>
<li>The 99% CI will be wider than the 95% CI</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="summary---hypothesis-testing">Summary - hypothesis testing<a class="anchor" aria-label="anchor" href="#summary---hypothesis-testing"></a>
</h2>
<div class="section level3">
<h3 id="power-and-type-i-and-ii-error">Power and type I and II error<a class="anchor" aria-label="anchor" href="#power-and-type-i-and-ii-error"></a>
</h3>
<table class="table">
<colgroup>
<col width="22%">
<col width="28%">
<col width="49%">
</colgroup>
<thead><tr class="header">
<th>True state of nature</th>
<th>Result of test</th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td></td>
<td><strong>Reject <span class="math inline">\(H_0\)</span></strong></td>
<td><strong>Fail to reject <span class="math inline">\(H_0\)</span></strong></td>
</tr>
<tr class="even">
<td>
<span class="math inline">\(H_0\)</span> TRUE</td>
<td>Type I error, probability = <span class="math inline">\(\alpha\)</span>
</td>
<td>No error, probability = <span class="math inline">\(1-\alpha\)</span>
</td>
</tr>
<tr class="odd">
<td>
<span class="math inline">\(H_0\)</span> FALSE</td>
<td>No error, probability is called power = <span class="math inline">\(1-\beta\)</span>
</td>
<td>Type II error, probability = <span class="math inline">\(\beta\)</span> (false negative)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="use-and-mis-use-of-the-p-value">Use and mis-use of the p-value<a class="anchor" aria-label="anchor" href="#use-and-mis-use-of-the-p-value"></a>
</h3>
<ul>
<li>The p-value is the probability of observing a sample statistic
<em>as or more extreme</em> as you did, <em>assuming that <span class="math inline">\(H_0\)</span> is true</em>
</li>
<li>The p-value is a <strong>random variable</strong>:
<ul>
<li>
<strong>don’t</strong> treat it as precise.</li>
<li>
<strong>don’t</strong> do silly things like try to interpret
differences or ratios between p-values</li>
<li>
<strong>don’t</strong> lose track of test assumptions such as
independence of observations</li>
<li>
<strong>do</strong> use a moderate p-value cutoff, then use some
effect size measure for ranking</li>
</ul>
</li>
<li>Small p-values are particularly:
<ul>
<li>variable under repeated sampling, and</li>
<li>sensitive to test assumptions</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="use-and-mis-use-of-the-p-value-contd">Use and mis-use of the p-value (cont’d)<a class="anchor" aria-label="anchor" href="#use-and-mis-use-of-the-p-value-contd"></a>
</h3>
<ul>
<li>If we fail to reject <span class="math inline">\(H_0\)</span>, is
the probability that <span class="math inline">\(H_0\)</span> is true
equal to (<span class="math inline">\(1-\alpha\)</span>)? (Hint: NO NO
NO!)
<ul>
<li>Failing to reject <span class="math inline">\(H_0\)</span> does not
mean <span class="math inline">\(H_0\)</span> is true</li>
<li>“No evidence of difference <span class="math inline">\(\neq\)</span>
evidence of no difference”</li>
</ul>
</li>
<li>Statistical significance vs. practical significance
<ul>
<li>As sample size increases, point estimates such as the mean are more
precise</li>
<li>With large sample size, small differences in means may be
statistically significant but not practically significant</li>
</ul>
</li>
<li>Although <span class="math inline">\(\alpha = 0.05\)</span> is a
common cut-off for the p-value, there is no set border between
“significant” and “insignificant,” only increasingly strong evidence
against <span class="math inline">\(H_0\)</span> (in favor of <span class="math inline">\(H_A\)</span>) as the p-value gets smaller.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="r---basic-usage">R - basic usage<a class="anchor" aria-label="anchor" href="#r---basic-usage"></a>
</h2>
<div class="section level3">
<h3 id="tips-for-learning-r">Tips for learning R<a class="anchor" aria-label="anchor" href="#tips-for-learning-r"></a>
</h3>
<table class="table">
<colgroup>
<col width="69%">
<col width="30%">
</colgroup>
<thead><tr class="header">
<th>Pseudo code</th>
<th>Example code</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>library(packagename)</td>
<td>library(dplyr)</td>
</tr>
<tr class="even">
<td>?functionname</td>
<td>?select</td>
</tr>
<tr class="odd">
<td>?package::functionname</td>
<td>?dplyr::select</td>
</tr>
<tr class="even">
<td>? ‘Reserved keyword or symbol’ </td>
<td>? ‘%&gt;%’</td>
</tr>
<tr class="odd">
<td>??searchforpossiblyexistingfunctionandortopic</td>
<td>??simulate</td>
</tr>
<tr class="even">
<td>help(package = “loadedpackage”)</td>
<td>help(“dplyr”)</td>
</tr>
<tr class="odd">
<td>browseVignettes(“packagename”)</td>
<td>browseVignettes(“dplyr”)</td>
</tr>
</tbody>
</table>
<p>Slide credit: Marcel Ramos</p>
</div>
<div class="section level3">
<h3 id="installing-packages-the-bioconductor-way">Installing Packages the Bioconductor Way<a class="anchor" aria-label="anchor" href="#installing-packages-the-bioconductor-way"></a>
</h3>
<ul>
<li>See the <a href="http://www.bioconductor.org/install" class="external-link">Bioconductor</a> site for
more info</li>
</ul>
<p>Pseudo code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span> <span class="co">#from CRAN</span></span>
<span><span class="va">packages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"packagename"</span>, <span class="st">"githubuser/repository"</span>, <span class="st">"biopackage"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">packages</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/valid.html" class="external-link">valid</a></span><span class="op">(</span><span class="op">)</span>  <span class="co">#check validity of installed packages</span></span></code></pre></div>
<ul>
<li>Works for CRAN, GitHub, and Bioconductor packages!</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="introduction-to-the-rbioconductor-data-classes">Introduction to the R/Bioconductor data classes<a class="anchor" aria-label="anchor" href="#introduction-to-the-rbioconductor-data-classes"></a>
</h2>
<div class="section level3">
<h3 id="base-r-data-types-atomic-vectors">Base R Data Types: atomic vectors<a class="anchor" aria-label="anchor" href="#base-r-data-types-atomic-vectors"></a>
</h3>
<p><code>numeric</code> (set seed to sync random number generator):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] -0.6264538  0.1836433 -0.8356286  1.5952808  0.3295078</span></span></code></pre></div>
<p><code>integer</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1L</span><span class="op">:</span><span class="fl">5L</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3 5 1 4 2</span></span></code></pre></div>
<p><code>logical</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">3</span></span>
<span><span class="co">#&gt; [1] FALSE FALSE  TRUE</span></span></code></pre></div>
<p><code>character</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yes"</span>, <span class="st">"no"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "yes" "no"</span></span></code></pre></div>
<p><code>factor</code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yes"</span>, <span class="st">"no"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] yes no </span></span>
<span><span class="co">#&gt; Levels: no yes</span></span></code></pre></div>
<p>For real-life recoding of factors, try <code><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">dplyr::recode</a></code>,
<code><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">dplyr::recode_factor</a></code></p>
</div>
<div class="section level3">
<h3 id="base-r-data-types-missingness">Base R Data Types: missingness<a class="anchor" aria-label="anchor" href="#base-r-data-types-missingness"></a>
</h3>
<ul>
<li>Missing Values and others - <strong>IMPORTANT</strong>
</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="cn">NaN</span>, <span class="op">-</span><span class="cn">Inf</span>, <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]   NA  NaN -Inf  Inf</span></span></code></pre></div>
<p><code><a href="https://rdrr.io/r/base/class.html" class="external-link">class()</a></code> to find the class of a variable.</p>
</div>
<div class="section level3">
<h3 id="base-r-data-types-matrix-list-data-frame">Base R Data Types: matrix, list, data.frame<a class="anchor" aria-label="anchor" href="#base-r-data-types-matrix-list-data-frame"></a>
</h3>
<p><code>matrix</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3]</span></span>
<span><span class="co">#&gt; [1,]    1    4    7</span></span>
<span><span class="co">#&gt; [2,]    2    5    8</span></span>
<span><span class="co">#&gt; [3,]    3    6    9</span></span></code></pre></div>
<p>The <code>list</code> is a non-atomic vector:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">measurements</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.3</span>, <span class="fl">1.6</span>, <span class="fl">3.2</span>, <span class="fl">9.8</span>, <span class="fl">10.2</span><span class="op">)</span></span>
<span><span class="va">parents</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Parent1.name"</span>, <span class="st">"Parent2.name"</span><span class="op">)</span></span>
<span><span class="va">my.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">measurements</span>, <span class="va">parents</span><span class="op">)</span></span>
<span><span class="va">my.list</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1]  1.3  1.6  3.2  9.8 10.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] "Parent1.name" "Parent2.name"</span></span></code></pre></div>
<p>The <code>data.frame</code> has list-like and matrix-like
properties:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">11</span><span class="op">:</span><span class="fl">16</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">.2</span><span class="op">)</span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"one"</span>, <span class="st">"two"</span>, <span class="st">"three"</span>, <span class="st">"four"</span>, <span class="st">"five"</span>, <span class="st">"six"</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span></span>
<span><span class="va">my.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">z</span>, <span class="va">a</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="bioconductor-s4-vectors-dataframe">Bioconductor S4 vectors: DataFrame<a class="anchor" aria-label="anchor" href="#bioconductor-s4-vectors-dataframe"></a>
</h3>
<ul>
<li>Bioconductor (www.bioconductor.org) defines its own set of vectors
using the S4 formal class system <code>DqtqFrame</code>: like a
<code>data.frame</code> but more flexible. columns can be any atomic
vector type:
<ul>
<li>
<code>GenomicRanges</code> objects</li>
<li>
<code>Rle</code> (run-length encoding)</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/S4Vectors" class="external-link">S4Vectors</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>var1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"b"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                var2 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>parent <span class="op">=</span> <span class="st">"DataFrame is my virtual parent"</span><span class="op">)</span></span>
<span><span class="va">df</span></span>
<span><span class="co">#&gt; DataFrame with 3 rows and 2 columns</span></span>
<span><span class="co">#&gt;    var1      var2</span></span>
<span><span class="co">#&gt;   &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1     a         1</span></span>
<span><span class="co">#&gt; 2     a         2</span></span>
<span><span class="co">#&gt; 3     b         3</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dataframe-is-actually-a-virtual-classes">DataFrame is actually a virtual classes<a class="anchor" aria-label="anchor" href="#dataframe-is-actually-a-virtual-classes"></a>
</h3>
<div class="grViz html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-69b5ed4906484ed42468" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-69b5ed4906484ed42468">{"x":{"diagram":"                           # All instructions are within a large character string\ndigraph surveillance_diagram {    # \"digraph\" means \"directional graph\", then the graph name \n  \n  # graph statement\n  #################\n  graph [layout = dot,\n         rankdir = TB,\n         overlap = true]\n  \n  # nodes\n  #######\n  node [shape = none,           # shape = circle\n       fixedsize = false,\n       fontsize = 10]               # width of circles\n  \n  DataFrame                         # names of nodes\n  DFrame\n  SQLDataFrame\n  ParquetDataFrame\n  \"...\"\n\n  # edges\n  #######\n  DFrame -> DataFrame \n  DFrame -> SQLDataFrame\n  DFrame -> ParquetDataFrame\n  DFrame -> \"...\"\n\n}\n","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><ul>
<li>Rationale: Methods defined on <code>DFrame</code> are shared by all
classes inheriting it</li>
<li>See: <a href="https://bioconductor.org/help/course-materials/2019/BiocDevelForum/02-DataFrame.pdf" class="external-link uri">https://bioconductor.org/help/course-materials/2019/BiocDevelForum/02-DataFrame.pdf</a>
</li>
<li>See also: <a href="https://github.com/Bioconductor/S4Vectors/issues/90#issue-1026425148" class="external-link uri">https://github.com/Bioconductor/S4Vectors/issues/90#issue-1026425148</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="list-and-derived-classes">List and derived classes<a class="anchor" aria-label="anchor" href="#list-and-derived-classes"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-class.html" class="external-link">List</a></span><span class="op">(</span><span class="va">my.list</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/List-class.html" class="external-link">List</a></span><span class="op">(</span><span class="va">my.list</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Formal class 'SimpleList' [package "S4Vectors"] with 4 slots</span></span>
<span><span class="co">#&gt;   ..@ listData       :List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:5] 1.3 1.6 3.2 9.8 10.2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:2] "Parent1.name" "Parent2.name"</span></span>
<span><span class="co">#&gt;   ..@ elementType    : chr "ANY"</span></span>
<span><span class="co">#&gt;   ..@ elementMetadata: NULL</span></span>
<span><span class="co">#&gt;   ..@ metadata       : list()</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/IRanges" class="external-link">IRanges</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">IntegerList</a></span><span class="op">(</span>var1 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">26</span>, var2 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt; IntegerList of length 2</span></span>
<span><span class="co">#&gt; [["var1"]] 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26</span></span>
<span><span class="co">#&gt; [["var2"]] 1 2 3 4 5 6 7 8 9 10 11 12 ... 89 90 91 92 93 94 95 96 97 98 99 100</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">CharacterList</a></span><span class="op">(</span>var1 <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>, var2 <span class="op">=</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">26</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; CharacterList of length 2</span></span>
<span><span class="co">#&gt; [["var1"]] a b c d e f g h i j ... &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; &lt;NA&gt;</span></span>
<span><span class="co">#&gt; [["var2"]] A B C D E F G H I J K L M N O P Q R S T U V W X Y Z</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">LogicalList</a></span><span class="op">(</span>var1 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fl">5</span>, var2 <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; LogicalList of length 2</span></span>
<span><span class="co">#&gt; [["var1"]] FALSE FALSE FALSE FALSE TRUE FALSE ... FALSE FALSE FALSE FALSE FALSE</span></span>
<span><span class="co">#&gt; [["var2"]] TRUE FALSE TRUE FALSE TRUE FALSE ... FALSE TRUE FALSE TRUE FALSE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="biostrings">Biostrings<a class="anchor" aria-label="anchor" href="#biostrings"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bstring</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XString-class.html" class="external-link">BString</a></span><span class="op">(</span><span class="st">"I am a BString object"</span><span class="op">)</span></span>
<span><span class="va">bstring</span></span>
<span><span class="co">#&gt; 21-letter BString object</span></span>
<span><span class="co">#&gt; seq: I am a BString object</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dnastring</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="st">"TTGAAA-CTC-N"</span><span class="op">)</span></span>
<span><span class="va">dnastring</span></span>
<span><span class="co">#&gt; 12-letter DNAString object</span></span>
<span><span class="co">#&gt; seq: <span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">AAA</span>-<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span>-<span style="color: #EEEEEE; background-color: #B2B2B2;">N</span></span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">dnastring</span><span class="op">)</span></span>
<span><span class="co">#&gt; Formal class 'DNAString' [package "Biostrings"] with 5 slots</span></span>
<span><span class="co">#&gt;   ..@ shared         :Formal class 'SharedRaw' [package "XVector"] with 2 slots</span></span>
<span><span class="co">#&gt;   .. .. ..@ xp                    :&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;   .. .. ..@ .link_to_cached_object:&lt;environment: 0x5584e9cf1480&gt; </span></span>
<span><span class="co">#&gt;   ..@ offset         : int 0</span></span>
<span><span class="co">#&gt;   ..@ length         : int 12</span></span>
<span><span class="co">#&gt;   ..@ elementMetadata: NULL</span></span>
<span><span class="co">#&gt;   ..@ metadata       : list()</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html" class="external-link">alphabetFrequency</a></span><span class="op">(</span><span class="va">dnastring</span>, baseOnly<span class="op">=</span><span class="cn">TRUE</span>, as.prob<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;          A          C          G          T      other </span></span>
<span><span class="co">#&gt; 0.25000000 0.16666667 0.08333333 0.25000000 0.25000000</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rangedsummarizedexperiment">(Ranged)SummarizedExperiment<a class="anchor" aria-label="anchor" href="#rangedsummarizedexperiment"></a>
</h3>
<p><img src="images/SE.svg" alt="Summarized Experiment"> * Source: <a href="https://bioconductor.org/packages/SummarizedExperiment/" class="external-link uri">https://bioconductor.org/packages/SummarizedExperiment/</a></p>
</div>
<div class="section level3">
<h3 id="rangedsummarizedexperiment-1">(Ranged)SummarizedExperiment<a class="anchor" aria-label="anchor" href="#rangedsummarizedexperiment-1"></a>
</h3>
<ul>
<li>
<code>RangedSummarizedExperiment</code> is the <em>de facto</em>
standard for “rectangular” data with metadata</li>
<li>Extended by <code>SingleCellExperiment</code>,
<code>DESeqDataSet</code>
</li>
<li>Emulated by <code>MultiAssayExperiment</code>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="lab-exercises-and-links">Lab Exercises and Links<a class="anchor" aria-label="anchor" href="#lab-exercises-and-links"></a>
</h3>
<ul>
<li>Lab Exercises
<ul>
<li><a href="http://bioconductor.org/books/3.17/OSCA.intro/getting-scrna-seq-datasets.html" class="external-link">OSCA
Introduction Chapter 3: Getting scRNA-seq datasets</a></li>
<li>
<a href="http://bioconductor.org/books/3.17/OSCA.intro/the-singlecellexperiment-class.html" class="external-link">OSCA
Introduction Chapter 4: The SingleCellExperiment class</a> (as time
allows)</li>
</ul>
</li>
<li>Links
<ul>
<li>A built <a href="http://waldronlab.io/AppStatBio/">html</a> version
of this lecture is available.</li>
<li>The <a href="https://github.com/waldronlab/AppStatBio" class="external-link">source</a> R
Markdown is also available from Github.</li>
</ul>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Levi Waldron.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
