<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="AppStatBio">
<title>Applied Statistics for High-throughput Biology: Session 2 • AppStatBio</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Applied Statistics for High-throughput Biology: Session 2">
<meta property="og:description" content="AppStatBio">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">AppStatBio</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/day1_intro.html">Applied Statistics for High-throughput Biology: Session 1</a>
    <a class="dropdown-item" href="../articles/day2_unsupervised.html">Applied Statistics for High-throughput Biology: Session 2</a>
    <a class="dropdown-item" href="../articles/day3_linearmodels.html">Applied Statistics for High-throughput Biology: Session 3</a>
    <a class="dropdown-item" href="../articles/day4_batcheffects-vis.html">Applied Statistics for High-throughput Biology: Session 4</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Applied Statistics for High-throughput Biology: Session 2</h1>
                        <h4 data-toc-skip class="author">Levi
Waldron</h4>
            
      
      
      <div class="d-none name"><code>day2_unsupervised.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="schematic-of-a-typical-scrna-seq-analysis-workflow">Schematic of a typical scRNA-seq analysis workflow<a class="anchor" aria-label="anchor" href="#schematic-of-a-typical-scrna-seq-analysis-workflow"></a>
</h2>
<div class="figure">
<img src="images/workflow.png" alt=""><p class="caption">Single-cell workflow</p>
</div>
<p>Each stage (separated by dashed lines) consists of a number of
specific steps, many of which operate on and modify a
SingleCellExperiment instance. (<a href="http://bioconductor.org/books/3.17/OSCA.intro/" class="external-link">original
image</a>)</p>
</div>
<div class="section level2">
<h2 id="day-4-outline">Day 4 outline<a class="anchor" aria-label="anchor" href="#day-4-outline"></a>
</h2>
<p><a href="https://genomicsclass.github.io/book/" class="external-link">Book</a> chapter
8:</p>
<ul>
<li>Distances in high dimensions</li>
<li>Principal Components Analysis and Singular Value Decomposition</li>
<li>Multidimensional Scaling</li>
<li>t-SNE and UMAP</li>
</ul>
</div>
<div class="section level2">
<h2 id="metrics-and-distances">Metrics and distances<a class="anchor" aria-label="anchor" href="#metrics-and-distances"></a>
</h2>
<p>A <strong>metric</strong> satisfies the following five
properties:</p>
<ol style="list-style-type: decimal">
<li>non-negativity <span class="math inline">\(d(a, b) \ge
0\)</span>
</li>
<li>symmetry <span class="math inline">\(d(a, b) = d(b, a)\)</span>
</li>
<li>identification mark <span class="math inline">\(d(a, a) =
0\)</span>
</li>
<li>definiteness <span class="math inline">\(d(a, b) = 0\)</span> if and
only if <span class="math inline">\(a=b\)</span>
</li>
<li>triangle inequality <span class="math inline">\(d(a, b) + d(b, c)
\ge d(a, c)\)</span>
</li>
</ol>
<ul>
<li>A <strong>distance</strong> is only required to satisfy 1-3.</li>
<li>A <strong>similarity function</strong> satisfies 1-2, and
<strong>increases</strong> as <span class="math inline">\(a\)</span> and
<span class="math inline">\(b\)</span> become more similar</li>
<li>A <strong>dissimilarity function</strong> satisfies 1-2, and
<strong>decreases</strong> as <span class="math inline">\(a\)</span> and
<span class="math inline">\(b\)</span> become more similar</li>
</ul>
</div>
<div class="section level2">
<h2 id="euclidian-distance-metric">Euclidian distance (metric)<a class="anchor" aria-label="anchor" href="#euclidian-distance-metric"></a>
</h2>
<ul>
<li>Remember grade school:
<img src="day2_unsupervised_files/figure-html/unnamed-chunk-2-1.png" width="700"><center>
Euclidean d = <span class="math inline">\(\sqrt{ (A_x-B_x)^2 +
(A_y-B_y)^2}\)</span>.
</center>
</li>
<li>
<strong>Side note</strong>: also referred to as <em><span class="math inline">\(L_2\)</span> norm</em>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="euclidian-distance-in-high-dimensions">Euclidian distance in high dimensions<a class="anchor" aria-label="anchor" href="#euclidian-distance-in-high-dimensions"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## BiocManager::install("genomicsclass/tissuesGeneExpression") #if needed</span></span>
<span><span class="co">## BiocManager::install("genomicsclass/GSE5859") #if needed</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSE5859</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tissuesGeneExpression</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tissuesGeneExpression</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="co">##gene expression data</span></span>
<span><span class="co">#&gt; [1] 22215   189</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span> <span class="co">##tissue[i] corresponds to e[,i]</span></span>
<span><span class="co">#&gt; tissue</span></span>
<span><span class="co">#&gt;  cerebellum       colon endometrium hippocampus      kidney       liver </span></span>
<span><span class="co">#&gt;          38          34          15          31          39          26 </span></span>
<span><span class="co">#&gt;    placenta </span></span>
<span><span class="co">#&gt;           6</span></span></code></pre></div>
<p>Interested in identifying similar <em>samples</em> and similar
<em>genes</em></p>
</div>
<div class="section level2">
<h2 id="notes-about-euclidian-distance-in-high-dimensions">Notes about Euclidian distance in high dimensions<a class="anchor" aria-label="anchor" href="#notes-about-euclidian-distance-in-high-dimensions"></a>
</h2>
<ul>
<li>Points are no longer on the Cartesian plane</li>
<li>instead they are in higher dimensions. For example:
<ul>
<li>sample <span class="math inline">\(i\)</span> is defined by a point
in 22,215 dimensional space: <span class="math inline">\((Y_{1,i},\dots,Y_{22215,i})^\top\)</span>.</li>
<li>feature <span class="math inline">\(g\)</span> is defined by a point
in 189 dimensions <span class="math inline">\((Y_{g,189},\dots,Y_{g,189})^\top\)</span>
</li>
</ul>
</li>
</ul>
<p>Euclidean distance as for two dimensions. E.g., the distance between
two samples <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> is:</p>
<p><span class="math display">\[ \mbox{dist}(i,j) = \sqrt{
\sum_{g=1}^{22215} (Y_{g,i}-Y_{g,j })^2 } \]</span></p>
<p>and the distance between two features <span class="math inline">\(h\)</span> and <span class="math inline">\(g\)</span> is:</p>
<p><span class="math display">\[ \mbox{dist}(h,g) = \sqrt{
\sum_{i=1}^{189} (Y_{h,i}-Y_{g,i})^2 } \]</span></p>
</div>
<div class="section level2">
<h2 id="euclidian-distance-in-matrix-algebra-notation">Euclidian distance in matrix algebra notation<a class="anchor" aria-label="anchor" href="#euclidian-distance-in-matrix-algebra-notation"></a>
</h2>
<p>The Euclidian distance between samples <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> can be written as:</p>
<p><span class="math display">\[ \mbox{dist}(i,j) = \sqrt{ (\mathbf{Y}_i
- \mathbf{Y}_j)^\top(\mathbf{Y}_i - \mathbf{Y}_j) }\]</span></p>
<p>with <span class="math inline">\(\mathbf{Y}_i\)</span> and <span class="math inline">\(\mathbf{Y}_j\)</span> columns <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3]</span></span>
<span><span class="co">#&gt; [1,]    1    2    3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1]</span></span>
<span><span class="co">#&gt; [1,]    1</span></span>
<span><span class="co">#&gt; [2,]    2</span></span>
<span><span class="co">#&gt; [3,]    3</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1]</span></span>
<span><span class="co">#&gt; [1,]   14</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="note-about-matrix-algebra-in-r">Note about matrix algebra in R<a class="anchor" aria-label="anchor" href="#note-about-matrix-algebra-in-r"></a>
</h2>
<ul>
<li>R is very efficient at matrix algebra</li>
<li>for very large matricies, see the:
<ul>
<li>
<a href="https://CRAN.R-project.org/package=Matrix" class="external-link">Matrix</a> CRAN
package (sparse matrices)</li>
<li>
<a href="https://bioconductor.org/packages/rhdf5/" class="external-link">rhdf5</a> and <a href="https://bioconductor.org/packages/DelayedArray/" class="external-link">DelayedArray</a>
Bioconductor package (on-disk arrays)</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="sample-example">3 sample example<a class="anchor" aria-label="anchor" href="#sample-example"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kidney1</span> <span class="op">&lt;-</span> <span class="va">e</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">kidney2</span> <span class="op">&lt;-</span> <span class="va">e</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">colon1</span> <span class="op">&lt;-</span> <span class="va">e</span><span class="op">[</span>, <span class="fl">87</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="va">kidney1</span> <span class="op">-</span> <span class="va">kidney2</span><span class="op">)</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 85.8546</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="va">kidney1</span> <span class="op">-</span> <span class="va">colon1</span><span class="op">)</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 122.8919</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sample-example-using-dist">3 sample example using dist()<a class="anchor" aria-label="anchor" href="#sample-example-using-dist"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 22215   189</span></span>
<span><span class="op">(</span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">e</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">87</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 GSM11805.CEL.gz GSM11814.CEL.gz</span></span>
<span><span class="co">#&gt; GSM11814.CEL.gz         85.8546                </span></span>
<span><span class="co">#&gt; GSM92240.CEL.gz        122.8919        115.4773</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "dist"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-dist-function">The dist() function<a class="anchor" aria-label="anchor" href="#the-dist-function"></a>
</h2>
<p>Excerpt from ?dist:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span></span>
<span>    <span class="va">x</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>    diag <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    upper <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    p <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li>
<strong>method:</strong> the distance measure to be used.
<ul>
<li>This must be one of “euclidean”, “maximum”, “manhattan”, “canberra”,
“binary” or “minkowski”. Any unambiguous substring can be given.</li>
</ul>
</li>
<li>
<code>dist</code> class output from <code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist()</a></code> is used for
many clustering algorithms and heatmap functions</li>
</ul>
<p><em>Caution</em>: <code>dist(e)</code> creates a 22215 x 22215 matrix
that will probably crash your R session.</p>
</div>
<div class="section level2">
<h2 id="note-on-standardization">Note on standardization<a class="anchor" aria-label="anchor" href="#note-on-standardization"></a>
</h2>
<ul>
<li>In practice, features (e.g. genes) are typically “standardized”,
<em>i.e.</em> converted to z-score:</li>
</ul>
<p><span class="math display">\[x_{gi} \leftarrow \frac{(x_{gi} -
\bar{x}_g)}{s_g}\]</span></p>
<ul>
<li>This is done because the differences in overall levels between
features are often not due to biological effects but technical ones,
<em>e.g.</em>:
<ul>
<li>GC bias, PCR amplification efficiency, …</li>
</ul>
</li>
<li>Euclidian distance and <span class="math inline">\(1-r\)</span>
(Pearson cor) are related:
<ul>
<li><span class="math inline">\(\frac{d_E(x, y)^2}{2m} = 1 -
r_{xy}\)</span></li>
<li>
<span class="math inline">\(m\)</span> = # dimensions</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="dimension-reduction-and-pca">Dimension reduction and PCA<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-pca"></a>
</h2>
<ul>
<li>Motivation for dimension reduction</li>
</ul>
<p>Simulate the heights of twin pairs:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/mvrnorm.html" class="external-link">mvrnorm</a></span><span class="op">(</span><span class="va">n</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.95</span>, <span class="fl">0.95</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]   2 100</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 1.0000000 0.9433295</span></span>
<span><span class="co">#&gt; [2,] 0.9433295 1.0000000</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualizing-twin-pairs-data">Visualizing twin pairs data<a class="anchor" aria-label="anchor" href="#visualizing-twin-pairs-data"></a>
</h2>
<p><img src="day2_unsupervised_files/figure-html/vistwins-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="not-much-distance-is-lost-in-the-second-dimension">Not much distance is lost in the second dimension<a class="anchor" aria-label="anchor" href="#not-much-distance-is-lost-in-the-second-dimension"></a>
</h2>
<p><img src="day2_unsupervised_files/figure-html/twindists-1.png" width="700"></p>
<ul>
<li>Not much loss of height differences when just using average heights
of twin pairs.
<ul>
<li>because twin heights are highly correlated</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="singular-value-decomposition-svd">Singular Value Decomposition (SVD)<a class="anchor" aria-label="anchor" href="#singular-value-decomposition-svd"></a>
</h2>
<p>SVD generalizes the example rotation we looked at:</p>
<p><span class="math display">\[\mathbf{Y} =
\mathbf{UDV}^\top\]</span></p>
<center>
<img src="images/SVD1.png" alt="SVD" align="middle" style="height: 350px;">
</center>
<ul>
<li><p><strong>note</strong>: the above formulation is for <span class="math inline">\(m\)</span> rows <span class="math inline">\(&gt;
n\)</span> columns</p></li>
<li><p><span class="math inline">\(\mathbf{Y}\)</span>: the <span class="math inline">\(m\)</span> rows x <span class="math inline">\(n\)</span> cols matrix of measurements</p></li>
<li><p><span class="math inline">\(\mathbf{U}\)</span>: <span class="math inline">\(m \times n\)</span> matrix relating original
scores to PCA scores (<strong>loadings</strong>)</p></li>
<li><p><span class="math inline">\(\mathbf{D}\)</span>: <span class="math inline">\(n \times n\)</span> diagonal matrix
(<strong>eigenvalues</strong>)</p></li>
<li>
<p><span class="math inline">\(\mathbf{V}\)</span>: <span class="math inline">\(n \times n\)</span> <em>orthogonal</em> matrix
(<strong>eigenvectors or PCA scores</strong>)</p>
<ul>
<li>orthogonal = unit length and “perpendicular” in 3-D</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="svd-of-gene-expression-dataset">SVD of gene expression dataset<a class="anchor" aria-label="anchor" href="#svd-of-gene-expression-dataset"></a>
</h2>
<p>Scaling:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e.standardize</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span>, scale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>SVD:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/svd.html" class="external-link">svd</a></span><span class="op">(</span><span class="va">e.standardize</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "d" "u" "v"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="components-of-svd-results">Components of SVD results<a class="anchor" aria-label="anchor" href="#components-of-svd-results"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">s</span><span class="op">$</span><span class="va">u</span><span class="op">)</span>     <span class="co"># loadings</span></span>
<span><span class="co">#&gt; [1] 22215   189</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">s</span><span class="op">$</span><span class="va">d</span><span class="op">)</span>  <span class="co"># eigenvalues</span></span>
<span><span class="co">#&gt; [1] 189</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">s</span><span class="op">$</span><span class="va">v</span><span class="op">)</span>     <span class="co"># d %*% vT = scores</span></span>
<span><span class="co">#&gt; [1] 189 189</span></span></code></pre></div>
<center>
<img src="images/SVD1.png" alt="SVD" align="middle" style="height: 200px;">
</center>
</div>
<div class="section level2">
<h2 id="pca-is-a-svd">PCA is a SVD<a class="anchor" aria-label="anchor" href="#pca-is-a-svd"></a>
</h2>
<ul>
<li>gene expression dataset</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rafalib</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rafalib/man/mypar.html" class="external-link">mypar</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">e.standardize</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">s</span><span class="op">$</span><span class="va">u</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">~</span> <span class="va">p</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/pcavssvd-1.png" width="700"></p>
<p><strong>Lesson:</strong> u and v can each be multiplied by -1
arbitrarily</p>
</div>
<div class="section level2">
<h2 id="pca-interpretation-loadings">PCA interpretation: loadings<a class="anchor" aria-label="anchor" href="#pca-interpretation-loadings"></a>
</h2>
<center>
<img src="images/SVD1.png" alt="SVD" align="middle" style="height: 225px;">
</center>
<ul>
<li>
<span class="math inline">\(\mathbf{U}\)</span>
(<strong>loadings</strong>): relate the <em>principal component</em>
axes to the original variables
<ul>
<li>think of principal component axes as a weighted combination of
original axes</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="visualizing-pca-loadings">Visualizing PCA loadings<a class="anchor" aria-label="anchor" href="#visualizing-pca-loadings"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">p</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Index of genes"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"Loadings of PC1"</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"PC1 loadings of each gene"</span><span class="op">)</span> <span class="co">#or, predict(p)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.03</span>, <span class="fl">0.03</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/visloadings-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="genes-with-high-pc1-loadings">Genes with high PC1 loadings<a class="anchor" aria-label="anchor" href="#genes-with-high-pc1-loadings"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e.pc1genes</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">e.standardize</span><span class="op">[</span><span class="va">p</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.03</span> <span class="op">|</span></span>
<span>                           <span class="va">p</span><span class="op">$</span><span class="va">rotation</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0.03</span>,<span class="op">]</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">e.pc1genes</span>,</span>
<span>    scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    show_rownames <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    show_colnames <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/pheatmaploadings-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="pca-interpretation-eigenvalues">PCA interpretation: eigenvalues<a class="anchor" aria-label="anchor" href="#pca-interpretation-eigenvalues"></a>
</h2>
<ul>
<li>
<span class="math inline">\(\mathbf{D}\)</span>
(<strong>eigenvalues</strong>): standard deviation scaling factor that
each decomposed variable is multiplied by.</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rafalib</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rafalib/man/mypar.html" class="external-link">mypar</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>    <span class="va">p</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">p</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">150</span><span class="op">)</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"% variance explained"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Screeplot"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/screeplot-1.png" width="480" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="pca-interpretation-eigenvalues-1">PCA interpretation: eigenvalues<a class="anchor" aria-label="anchor" href="#pca-interpretation-eigenvalues-1"></a>
</h2>
<p>Alternatively as cumulative % variance explained (using
<code><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum()</a></code> function)</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rafalib</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rafalib/man/mypar.html" class="external-link">mypar</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="va">p</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">p</span><span class="op">$</span><span class="va">sdev</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"cumulative % variance explained"</span>,</span>
<span>    ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Cumulative screeplot"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/cumscreeplot-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="pca-interpretation-scores">PCA interpretation: scores<a class="anchor" aria-label="anchor" href="#pca-interpretation-scores"></a>
</h2>
<center>
<img src="images/SVD1.png" alt="SVD" align="middle" style="height: 225px;">
</center>
<ul>
<li>
<span class="math inline">\(\mathbf{V}\)</span>
(<strong>scores</strong>): The “datapoints” in the reduced prinipal
component space</li>
<li>In some implementations (like <code><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp()</a></code>), scores are
already scaled by eigenvalues: <span class="math inline">\(\mathbf{D
V^T}\)</span>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="pca-interpretation-scores-1">PCA interpretation: scores<a class="anchor" aria-label="anchor" href="#pca-interpretation-scores-1"></a>
</h2>
<p><img src="day2_unsupervised_files/figure-html/scores-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="multi-dimensional-scaling-mds">Multi-dimensional Scaling (MDS)<a class="anchor" aria-label="anchor" href="#multi-dimensional-scaling-mds"></a>
</h2>
<ul>
<li>also referred to as Principal Coordinates Analysis (PCoA)</li>
<li>a reduced SVD, performed on a distance matrix</li>
<li>identify two (or more) eigenvalues/vectors that preserve
distances</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">e.standardize</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html" class="external-link">cmdscale</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/plotmds-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="t-sne">t-SNE<a class="anchor" aria-label="anchor" href="#t-sne"></a>
</h2>
<ul>
<li>non-linear dimension reduction method very popular for visualizing
single-cell data
<ul>
<li>almost magical ability to show clearly separated clusters</li>
<li>performs different transformations on different regions</li>
</ul>
</li>
<li>computationally intensive so usually done only on top ~30 PCs</li>
<li>t-SNE is sensitive to choices of tuning parameters
<ul>
<li>“perplexity” parameter defines (loosely) how to balance attention
between local and global aspects of data</li>
<li>optimal choice of perplexity changes for different numbers of cells
from the <strong>same</strong> sample.</li>
<li>perplexity = <span class="math inline">\(\sqrt{N}\)</span> is one
rule of thumb. <span class="math inline">\(max(N/5, 50)\)</span> is
another (default of <a href="https://cran.r-project.org/package=Rtsne" class="external-link">Rtsne</a>)</li>
<li>Here is a <a href="https://towardsdatascience.com/how-to-tune-hyperparameters-of-tsne-7c0596a18868" class="external-link">good
post by Nikolay Oskolkov</a> on this topic.</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="t-sne-caveats">t-SNE caveats<a class="anchor" aria-label="anchor" href="#t-sne-caveats"></a>
</h2>
<ul>
<li>uses a random number generator</li>
<li>apparent spread of clusters is completely meaningless</li>
<li>distance between clusters might also not mean anything</li>
<li>parameters can be tuned to make data appear how you want</li>
<li>can show apparent clusters in random noise. Should <em>not</em> be
used for statistical inference</li>
<li>Try it to gain some intuition: <a href="https://distill.pub/2016/misread-tsne/" class="external-link uri">https://distill.pub/2016/misread-tsne/</a>)</li>
</ul>
<div class="figure">
<img src="images/tSNE-eg.png" alt=""><p class="caption">tSNE</p>
</div>
</div>
<div class="section level2">
<h2 id="pca-of-zeisel-single-cell-rna-seq-dataset">PCA of Zeisel single-cell RNA-seq dataset<a class="anchor" aria-label="anchor" href="#pca-of-zeisel-single-cell-rna-seq-dataset"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.zeisel</span> <span class="op">&lt;-</span> <span class="fu">fixedPCA</span><span class="op">(</span><span class="va">sce.zeisel</span>, subset.row<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce.zeisel</span>, dimred<span class="op">=</span><span class="st">"PCA"</span>, colour_by<span class="op">=</span><span class="st">"level1class"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="day2_unsupervised_files/figure-html/plotpcazeisel-1.png" alt="Principal Components Analysis of Zeisel dataset" width="700"><p class="caption">
Principal Components Analysis of Zeisel dataset
</p>
</div>
</div>
<div class="section level2">
<h2 id="t-sne-of-the-same-dataset">t-SNE of the same dataset<a class="anchor" aria-label="anchor" href="#t-sne-of-the-same-dataset"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.zeisel</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sce.zeisel</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce.zeisel</span>, dimred<span class="op">=</span><span class="st">"TSNE"</span>, colour_by<span class="op">=</span><span class="st">"level1class"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="day2_unsupervised_files/figure-html/tsenzeisel-1.png" alt="t-SNE clustering of Zeisel dataset" width="700"><p class="caption">
t-SNE clustering of Zeisel dataset
</p>
</div>
</div>
<div class="section level2">
<h2 id="umap-vs-t-sne">UMAP vs t-SNE<a class="anchor" aria-label="anchor" href="#umap-vs-t-sne"></a>
</h2>
<ul>
<li>UMAP may better preserve local and global distances</li>
<li>tends to have more compact visual clusters with more empty space
between them</li>
<li>more computationally efficient</li>
<li>also involves random number generation</li>
<li>Note: I prefer <em>not</em> setting the random number seed during
exploratory analysis in order to see the random variability</li>
</ul>
</div>
<div class="section level2">
<h2 id="umap-of-the-same-dataset">UMAP of the same dataset<a class="anchor" aria-label="anchor" href="#umap-of-the-same-dataset"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce.zeisel</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">sce.zeisel</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sce.zeisel</span>, dimred<span class="op">=</span><span class="st">"UMAP"</span>, colour_by<span class="op">=</span><span class="st">"level1class"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="day2_unsupervised_files/figure-html/umapzeisel-1.png" alt="UMAP representation of the Zeisel dataset" width="700"><p class="caption">
UMAP representation of the Zeisel dataset
</p>
</div>
</div>
<div class="section level2">
<h2 id="tsne-of-tissue-microarray-data">tSNE of tissue microarray data<a class="anchor" aria-label="anchor" href="#tsne-of-tissue-microarray-data"></a>
</h2>
<p>Using default parameters and no log transformation</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sct</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span>
<span><span class="va">sct</span><span class="op">$</span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="va">tissue</span></span>
<span><span class="co"># fake the log normalization because it is undesirable for microarray data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">assays</span><span class="op">(</span><span class="va">sct</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"logcounts"</span> </span>
<span><span class="va">sct</span> <span class="op">&lt;-</span> <span class="fu">fixedPCA</span><span class="op">(</span><span class="va">sct</span>, subset.row<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">sct</span> <span class="op">&lt;-</span> <span class="fu">runTSNE</span><span class="op">(</span><span class="va">sct</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sct</span>, dimred<span class="op">=</span><span class="st">"TSNE"</span>, colour_by<span class="op">=</span><span class="st">"tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/umaptsnezeisel-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="umap-of-tissue-microarray-data">UMAP of tissue microarray data<a class="anchor" aria-label="anchor" href="#umap-of-tissue-microarray-data"></a>
</h2>
<p>Also using default parameters and no log transformation</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sct</span> <span class="op">&lt;-</span> <span class="fu">runUMAP</span><span class="op">(</span><span class="va">sct</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu">plotReducedDim</span><span class="op">(</span><span class="va">sct</span>, dimred<span class="op">=</span><span class="st">"UMAP"</span>, colour_by<span class="op">=</span><span class="st">"tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="day2_unsupervised_files/figure-html/nologumap-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="summary-distances-and-dimension-reduction">Summary: distances and dimension reduction<a class="anchor" aria-label="anchor" href="#summary-distances-and-dimension-reduction"></a>
</h2>
<ul>
<li>
<strong>Note</strong>: signs of eigenvalues (square to get
variances) and eigenvectors (loadings) can be arbitrarily flipped</li>
<li>PCA and MDS are useful for dimension reduction when you have
<strong>correlated variables</strong>
</li>
<li>Variables are always centered.<br>
</li>
<li>Variables are also scaled unless you know they have the same scale
in the population</li>
<li>PCA projection can be applied to new datasets if you know the matrix
calculations</li>
<li>PCA is subject to over-fitting, screeplot can be tested by
cross-validation</li>
<li>PCA is often used prior to t-SNE and UMAP for de-noising and
computational tractability</li>
</ul>
</div>
<div class="section level2">
<h2 id="lab-exercise">Lab exercise<a class="anchor" aria-label="anchor" href="#lab-exercise"></a>
</h2>
<ul>
<li>OSCA Basics: <a href="http://bioconductor.org/books/3.17/OSCA.basic/dimensionality-reduction.html" class="external-link">Chapter
4 Dimensionality Reduction</a>
</li>
<li>Optional if you are interested, OSCA Advanced: <a href="http://bioconductor.org/books/3.17/OSCA.advanced/dimensionality-reduction-redux.html" class="external-link">Chapter
4 Dimensionality reduction, redux</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li>A built <a href="https://rpubs.com/lwaldron/AppStatBio2023_day2" class="external-link">html</a> version
of this lecture is available.</li>
<li>The <a href="https://github.com/waldronlab/AppStatBio" class="external-link">source</a> R
Markdown is also available from Github.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Levi Waldron.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
