<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="AppStatBio">
<title>Applied Statistics for High-throughput Biology: Session 4 • AppStatBio</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Applied Statistics for High-throughput Biology: Session 4">
<meta property="og:description" content="AppStatBio">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">AppStatBio</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/day1_intro.html">Applied Statistics for High-throughput Biology: Session 1</a>
    <a class="dropdown-item" href="../articles/day2_unsupervised.html">Applied Statistics for High-throughput Biology: Session 2</a>
    <a class="dropdown-item" href="../articles/day3_linearmodels.html">Applied Statistics for High-throughput Biology: Session 3</a>
    <a class="dropdown-item" href="../articles/day4_batcheffects-vis.html">Applied Statistics for High-throughput Biology: Session 4</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/waldronlab/AppStatBio/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Applied Statistics for High-throughput Biology: Session 4</h1>
                        <h4 data-toc-skip class="author">Levi
Waldron</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/waldronlab/AppStatBio/blob/HEAD/vignettes/day4_batcheffects-vis.Rmd" class="external-link"><code>vignettes/day4_batcheffects-vis.Rmd</code></a></small>
      <div class="d-none name"><code>day4_batcheffects-vis.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="outline">Outline<a class="anchor" aria-label="anchor" href="#outline"></a>
</h3>
<ul>
<li>
<a href="https://genomicsclass.github.io/book/pages/exploratory_data_analysis.html" class="external-link">Exploratory
data analysis</a> (Chapter 2)
<ul>
<li>Quantile Quantile (QQ)-plots</li>
<li>Boxplots</li>
<li>Scatterplots</li>
<li>Volcano plots</li>
<li>p-value histograms</li>
<li>MA plots</li>
<li>heatmaps</li>
</ul>
</li>
<li>
<a href="http://genomicsclass.github.io/book/pages/intro_to_batch_effects.html" class="external-link">Batch
effects</a> (Chapter 10)</li>
</ul>
</div>
<div class="section level2">
<h2 id="exploratory-data-analysis">Exploratory data analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h2>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<blockquote>
<p>“The greatest value of a picture is when it forces us to notice what
we never expected to see.” - John W. Tukey</p>
</blockquote>
<ul>
<li>Discover biases, systematic errors and unexpected variability in
data</li>
<li>Graphical approach to detecting these issues</li>
<li>Represents a first step in data analysis and guides hypothesis
testing</li>
<li>Opportunities for discovery in the outliers</li>
</ul>
</div>
<div class="section level3">
<h3 id="quantile-quantile-plots">Quantile Quantile Plots<a class="anchor" aria-label="anchor" href="#quantile-quantile-plots"></a>
</h3>
<ul>
<li>Quantiles divide a distribution into equally sized bins</li>
<li>Division into 100 bins gives percentiles</li>
<li>Quantiles of a theoretical distribution are plotted against an
experimental distribution
<ul>
<li>alternatively, quantiles of two experimental distributions</li>
</ul>
</li>
<li>Given a perfect fit, <span class="math inline">\(x=y\)</span>
</li>
<li>Useful in determining data distribution (normal, t, etc.)</li>
</ul>
</div>
<div class="section level3">
<h3 id="example-quantile-quantile-plots">Example: Quantile Quantile Plots<a class="anchor" aria-label="anchor" href="#example-quantile-quantile-plots"></a>
</h3>
<p><img src="day4_batcheffects-vis_files/figure-html/qq1-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="boxplots-about">Boxplots: About<a class="anchor" aria-label="anchor" href="#boxplots-about"></a>
</h3>
<ul>
<li>Provide a graph that is easy to interpret where data is not normally
distributed</li>
<li>Would be an appropriate choice to explore income data, as
distribution is highly skewed</li>
<li>Particularly informative in relation to outliers and range</li>
<li>Possible to compare multiple distributions side by side</li>
</ul>
</div>
<div class="section level3">
<h3 id="boxplots-example">Boxplots: Example<a class="anchor" aria-label="anchor" href="#boxplots-example"></a>
</h3>
<img src="day4_batcheffects-vis_files/figure-html/bp1-1.png" width="768"><center>
Three different views of a continuous variable
</center>
</div>
<div class="section level3">
<h3 id="scatterplots-and-correlation-about">Scatterplots And Correlation: About<a class="anchor" aria-label="anchor" href="#scatterplots-and-correlation-about"></a>
</h3>
<ul>
<li>For two continuous variables, scatter plot and calculation of
correlation is useful</li>
<li>Provides a graphical and numeric estimation of relationships</li>
<li>Quick and easy with <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> and <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor()</a></code>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="scatterplots-and-correlation-example">Scatterplots And Correlation: Example<a class="anchor" aria-label="anchor" href="#scatterplots-and-correlation-example"></a>
</h3>
<p><img src="day4_batcheffects-vis_files/figure-html/cor1-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="exploratory-data-analysis-in-high-dimensions">Exploratory data analysis in high dimensions<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis-in-high-dimensions"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span> <span class="co">##this loads three tables</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "matrix"     "array"      "data.frame"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2]</span></span>
<span><span class="co">#&gt; [1,] 8793   24</span></span>
<span><span class="co">#&gt; [2,]   24    4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">)</span></span>
<span><span class="co">#&gt;     ethnicity       date         filename group</span></span>
<span><span class="co">#&gt; 107       ASN 2005-06-23 GSM136508.CEL.gz     1</span></span>
<span><span class="co">#&gt; 122       ASN 2005-06-27 GSM136530.CEL.gz     1</span></span>
<span><span class="co">#&gt; 113       ASN 2005-06-27 GSM136517.CEL.gz     1</span></span>
<span><span class="co">#&gt; 163       ASN 2005-10-28 GSM136576.CEL.gz     1</span></span>
<span><span class="co">#&gt; 153       ASN 2005-10-07 GSM136566.CEL.gz     1</span></span>
<span><span class="co">#&gt; 161       ASN 2005-10-07 GSM136574.CEL.gz     1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="volcano-plots-setup">Volcano plots: Setup<a class="anchor" aria-label="anchor" href="#volcano-plots-setup"></a>
</h3>
<p>T-test for every row (gene) of gene expression matrix:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/genefilter/man/rowFtests.html" class="external-link">rowttests</a></span><span class="op">(</span><span class="va">geneExpression</span>, <span class="va">g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.006   0.000   0.006</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">p.value</span></span></code></pre></div>
<p>Note that these 8,793 tests are done in about 0.01s</p>
</div>
<div class="section level3">
<h3 id="volcano-plots-example">Volcano plots: Example<a class="anchor" aria-label="anchor" href="#volcano-plots-example"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">dm</span>,</span>
<span>     <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"Effect size (difference in group means)"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"- log (base 10) p-values"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottomright"</span>,</span>
<span>       lty <span class="op">=</span> <span class="fl">1</span>,</span>
<span>       col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>       legend <span class="op">=</span> <span class="st">"Bonferroni = 0.05"</span><span class="op">)</span></span></code></pre></div>
<p><img src="day4_batcheffects-vis_files/figure-html/vp2-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="volcano-plots-summary">Volcano plots: Summary<a class="anchor" aria-label="anchor" href="#volcano-plots-summary"></a>
</h3>
<ul>
<li>Many small p-values with small effect size indicate low within-group
variability</li>
<li>Inspect for asymmetry</li>
<li>Can color points by significance threshold</li>
</ul>
</div>
<div class="section level3">
<h3 id="p-value-histograms-setup">P-value histograms: Setup<a class="anchor" aria-label="anchor" href="#p-value-histograms-setup"></a>
</h3>
<ul>
<li>If all null hypotheses are true, expect a flat histogram of
p-values:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">randomData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span> <span class="op">*</span> <span class="va">m</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span>
<span><span class="va">nullpvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/genefilter/man/rowFtests.html" class="external-link">rowttests</a></span><span class="op">(</span><span class="va">randomData</span>, <span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="p-value-histograms-example">P-value histograms: Example<a class="anchor" aria-label="anchor" href="#p-value-histograms-example"></a>
</h3>
<p><img src="day4_batcheffects-vis_files/figure-html/pvalhist2-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="p-value-histograms-example-2-permutation">P-value histograms: Example 2 (permutation)<a class="anchor" aria-label="anchor" href="#p-value-histograms-example-2-permutation"></a>
</h3>
<p>Note that permuting these data doesn’t produce an ideal null p-value
histogram due to batch effects:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">permg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="va">permresults</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/genefilter/man/rowFtests.html" class="external-link">rowttests</a></span><span class="op">(</span><span class="va">geneExpression</span>, <span class="va">permg</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">permresults</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span></span></code></pre></div>
<p><img src="day4_batcheffects-vis_files/figure-html/pvalhist3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="p-value-histograms-summary">P-value histograms: Summary<a class="anchor" aria-label="anchor" href="#p-value-histograms-summary"></a>
</h3>
<ul>
<li>Give a quick look at how many significant p-values there may be</li>
<li>When using permuted labels, can exposes non-independence among the
samples
<ul>
<li>can be due to batch effects or family structure</li>
</ul>
</li>
<li>Most common approaches for correcting batch effects are:
<ul>
<li>
<code>ComBat</code>: corrects for known batch effects by linear
model), and</li>
<li>
<code>sva</code>: creates surrogate variables for unknown batch
effects, corrects the structure of permutation p-values</li>
<li>correction using control (housekeeping) genes</li>
<li>
<code>batchelor</code> for single-cell analysis</li>
</ul>
</li>
</ul>
<p><code>ComBat</code> and <code>sva</code> are available from the <a href="https://www.bioconductor.org/packages/sva" class="external-link">sva</a> Bioconductor
package</p>
</div>
<div class="section level3">
<h3 id="ma-plot">MA plot<a class="anchor" aria-label="anchor" href="#ma-plot"></a>
</h3>
<ul>
<li>just a scatterplot rotated 45<span class="math inline">\(^o\)</span>
</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rafalib</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rafalib/man/mypar.html" class="external-link">mypar</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">pseudo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">geneExpression</span>, <span class="fl">1</span>, <span class="va">median</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pseudo</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">pseudo</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span>, <span class="op">(</span><span class="va">geneExpression</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">-</span> <span class="va">pseudo</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="day4_batcheffects-vis_files/figure-html/pvalhist4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="ma-plot-summary">MA plot: Summary<a class="anchor" aria-label="anchor" href="#ma-plot-summary"></a>
</h3>
<ul>
<li>useful for quality control of high-dimensional data</li>
<li>plot all data values for a sample against another sample or a median
“pseudosample”</li>
<li>
<code>affyPLM::MAplots</code> better MA plots
<ul>
<li>adds a smoothing line to highlight departures from horizontal
line</li>
<li>plots a “cloud” rather than many data points</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<ul>
<li>Detailed representation of high-dimensional dataset.
<ul>
<li>
<code>ComplexHeatmap</code> package is the best as of 2023: large
datasets, interactive heatmaps, simple defaults but many customizations
possible</li>
</ul>
</li>
</ul>
<p><img src="day4_batcheffects-vis_files/figure-html/ma1-1.png" width="1152"></p>
</div>
<div class="section level3">
<h3 id="heatmaps-summary">Heatmaps: Summary<a class="anchor" aria-label="anchor" href="#heatmaps-summary"></a>
</h3>
<ul>
<li>Clustering becomes slow and memory-intensivefor thousands of rows
<ul>
<li>probably too detailed for thousands of rows</li>
</ul>
</li>
<li>can show co-expressed genes, groups of samples</li>
</ul>
</div>
<div class="section level3">
<h3 id="colors">Colors<a class="anchor" aria-label="anchor" href="#colors"></a>
</h3>
<ul>
<li>Types of color pallettes:
<ul>
<li>
<strong>sequential</strong>: shows a gradient</li>
<li>
<strong>diverging</strong>: goes in two directions from a
center</li>
<li>
<strong>qualitative</strong>: for categorical variables</li>
</ul>
</li>
<li>Keep color blindness in mind (10% of all men)</li>
</ul>
</div>
<div class="section level3">
<h3 id="colors-contd">Colors (cont’d)<a class="anchor" aria-label="anchor" href="#colors-contd"></a>
</h3>
<p>Combination of <code>RColorBrewer</code> package and
<code><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette()</a></code> can create anything you want</p>
<p><img src="day4_batcheffects-vis_files/figure-html/brewer-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plots-to-avoid">Plots To Avoid<a class="anchor" aria-label="anchor" href="#plots-to-avoid"></a>
</h3>
<blockquote>
<p>“Pie charts are a very bad way of displaying information.” - R
Help</p>
</blockquote>
<ul>
<li>Avoid pie charts</li>
<li>Avoid doughnut charts too</li>
<li>Avoid pseudo 3D</li>
<li>Use color judiciously</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="batch-effects">Batch effects<a class="anchor" aria-label="anchor" href="#batch-effects"></a>
</h2>
<div class="section level3">
<h3 id="pervasiveness-of-batch-effects">Pervasiveness of batch Effects<a class="anchor" aria-label="anchor" href="#pervasiveness-of-batch-effects"></a>
</h3>
<ul>
<li>pervasive in genomics (e.g. <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=20838408" class="external-link">Leek <em>et
al.</em> Nat Rev Genet. 2010 Oct;11(10):733-9.</a>)</li>
<li>affect DNA and RNA sequencing, proteomics, imaging, microarray…</li>
<li>have caused high-profile problems and retractions
<ul>
<li>you can’t get rid of them</li>
<li>but you can make sure they are not confounded with your
experiment</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="batch-effects---an-example">Batch Effects - an example<a class="anchor" aria-label="anchor" href="#batch-effects---an-example"></a>
</h3>
<ul>
<li>Nat Genet. 2007 Feb;39(2):226-31. Epub 2007 Jan 7.</li>
<li>Title: <em>Common genetic variants account for differences in gene
expression among ethnic groups.</em>
<ul>
<li>“The results show that specific genetic variation among populations
contributes appreciably to differences in gene expression
phenotypes.”</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSE5859</span><span class="op">)</span> <span class="co">## BiocInstaller::biocLite("genomicsclass/GSE5859")</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GSE5859</span><span class="op">)</span></span>
<span><span class="va">geneExpression</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span>
<span><span class="va">sampleInfo</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Note: the <code>ExpressionSet</code> object is obsolete, we use
<code>SummarizedExperiment</code> now</li>
</ul>
</div>
<div class="section level3">
<h3 id="date-of-processing-as-a-proxy-for-batch">Date of processing as a proxy for batch<a class="anchor" aria-label="anchor" href="#date-of-processing-as-a-proxy-for-batch"></a>
</h3>
<ul>
<li>Sample metadata included <em>date of processing</em>:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2002-10-31 2002-11-15 2002-11-19 2002-11-21 2002-11-22 2002-11-27 </span></span>
<span><span class="co">#&gt;          2          2          2          5          4          2</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">year</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, <span class="st">"%y"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">year</span> <span class="op">&lt;-</span> <span class="va">year</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">year</span><span class="op">)</span></span>
<span><span class="va">month</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, <span class="st">"%m"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">12</span> <span class="op">*</span> <span class="va">year</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">year</span>, <span class="va">sampleInfo</span><span class="op">$</span><span class="va">ethnicity</span><span class="op">)</span></span>
<span><span class="co">#&gt;     </span></span>
<span><span class="co">#&gt; year ASN CEU HAN</span></span>
<span><span class="co">#&gt;    0   0  32   0</span></span>
<span><span class="co">#&gt;    1   0  54   0</span></span>
<span><span class="co">#&gt;    2   0  13   0</span></span>
<span><span class="co">#&gt;    3  80   3   0</span></span>
<span><span class="co">#&gt;    4   2   0  24</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-batch-effects-by-pca">Visualizing batch effects by PCA<a class="anchor" aria-label="anchor" href="#visualizing-batch-effects-by-pca"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span>, scale. <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span></span>
<span>    <span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">~</span> <span class="va">month</span>,</span>
<span>    varwidth <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    notch <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"PC1 scores vs. month"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Month"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"PC1"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="day4_batcheffects-vis_files/figure-html/ge4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="visualizing-batch-effects-by-mds">Visualizing batch effects by MDS<a class="anchor" aria-label="anchor" href="#visualizing-batch-effects-by-mds"></a>
</h3>
<p>A starting point for a color palette:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">display.brewer.all</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span>, colorblindFriendly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="day4_batcheffects-vis_files/figure-html/ge5-1.png" width="700"></p>
<p>Interpolate one color per month on a quantitative palette:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">col3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span>, <span class="st">"Greys"</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="st">"black"</span><span class="op">)</span></span>
<span><span class="va">MYcols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="va">col3</span>, space <span class="op">=</span> <span class="st">"Lab"</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">month</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-batch-effects-by-mds-1">Visualizing batch effects by MDS<a class="anchor" aria-label="anchor" href="#visualizing-batch-effects-by-mds-1"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html" class="external-link">cmdscale</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mds</span>, col <span class="op">=</span> <span class="va">MYcols</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">month</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"MDS shaded by month"</span><span class="op">)</span></span></code></pre></div>
<p><img src="day4_batcheffects-vis_files/figure-html/mds1-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="the-batch-effects-impact-clustering">The batch effects impact clustering<a class="anchor" aria-label="anchor" href="#the-batch-effects-impact-clustering"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hcclass</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">hcclass</span>, <span class="va">year</span><span class="op">)</span></span>
<span><span class="co">#&gt;        year</span></span>
<span><span class="co">#&gt; hcclass  0  1  2  3  4</span></span>
<span><span class="co">#&gt;       1  0  2  8  1  0</span></span>
<span><span class="co">#&gt;       2 25 43  0  2  0</span></span>
<span><span class="co">#&gt;       3  6  0  0  0  0</span></span>
<span><span class="co">#&gt;       4  1  7  0  0  0</span></span>
<span><span class="co">#&gt;       5  0  2  5 80 26</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="approaches-to-correcting-for-batch-effects">Approaches to correcting for batch effects<a class="anchor" aria-label="anchor" href="#approaches-to-correcting-for-batch-effects"></a>
</h3>
<ul>
<li>
<em>No correction</em>
<ul>
<li>in my experience, the best choice for machine learning
applications</li>
</ul>
</li>
<li>
<a href="http://bioconductor.org/books/3.17/OSCA.multisample/integrating-datasets.html#by-rescaling-the-counts" class="external-link">Simple
rescaling</a>
<ul>
<li>Rescale observations (cells) to the same mean, variance in each
batch</li>
<li>maintains sparsity (ie zeros remain zeros)</li>
<li><code>batchelor::rescaleBatches()</code></li>
</ul>
</li>
<li>
<em>Linear modeling to regress out batch effects</em>
<ul>
<li>use when batches are known. Fits a linear model and use
residuals</li>
<li>assumes the same composition of cells across batches</li>
<li>
<code><a href="https://rdrr.io/pkg/limma/man/removeBatchEffect.html" class="external-link">limma::removeBatchEffect()</a></code>, <code>sva::comBat()</code>,
<code>batchelor::regressBatches()</code>
</li>
</ul>
</li>
<li>
<em>Linear modeling to achieve a flat p-value histogram when
permuting labels</em>
<ul>
<li>can be used when batches are unknown</li>
<li>“Surrogate Variables Analysis” implemented by the <a href="https://bioconductor.org/packages/sva/" class="external-link">sva</a> package</li>
</ul>
</li>
<li>
<em>Mutual Nearest Neighbors</em>
<ul>
<li>developed specifically for single-cell RNA-seq</li>
<li>no assumption of the same composition of cells across batches, but
still assumes no meaningful biological differences exist between
batches</li>
<li><code>batchelor::fastMNN()</code></li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="batch-effects---summary">Batch Effects - summary<a class="anchor" aria-label="anchor" href="#batch-effects---summary"></a>
</h3>
<ul>
<li>batches can be corrected for <strong>only</strong> if not
overlapping with conditions of interest
<ul>
<li>if confounded with treatment or outcome of interest, nothing can
help you</li>
<li>randomization of samples to batches in study design protects against
this</li>
</ul>
</li>
<li>during experimental design:
<ul>
<li>keep track of anything that might cause a batch effect for post-hoc
analysis</li>
<li>include control samples in each batch</li>
</ul>
</li>
<li>tend to affect many or all measurements by a little bit</li>
</ul>
</div>
<div class="section level3">
<h3 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h3>
<ul>
<li>OSCA Multi-sample <a href="http://bioconductor.org/books/3.17/OSCA.multisample/integrating-datasets.html" class="external-link">Chapter
1: Correcting batch effects</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h3>
<ul>
<li>A built <a href="https://rpubs.com/lwaldron/AppStatBio2023_day4" class="external-link">html</a> version
of this lecture is available.</li>
<li>The <a href="https://github.com/waldronlab/AppStatBio" class="external-link">source</a> R
Markdown is available from Github.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Levi Waldron.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
